File: javascript/leetcode/find-all-anagrams-in-a-string.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */

var findAnagrams = function(s, p) {
    foundIndexes = []
    pLen = p.length
    for (var i = 0; i < s.length - (pLen - 1); i++) {
      if (subStringIsAnagram(s.substring(i, i + pLen), p)) {
        foundIndexes.push(i)
      }
    }
    return foundIndexes;
};

var subStringIsAnagram = function (s, p) {
  pArr = p.split('')
  for (var c = 0; c < s.length; c++) {
    letterIndex = pArr.indexOf(s[c])
    if (letterIndex > -1) {
      pArr.splice(letterIndex, 1)
    } else {
      return false;
    }
  }
  return true
}

File: java/leetcode/find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams (String s, String p) {
        List<Integer> foundIndexes = new ArrayList();
        int pLen = p.length();
        for (int i = 0; i < s.length() - (pLen - 1); i++) {
            if (substringIsAnagram(s.substring(i, i + pLen), p)) {
                foundIndexes.add(i);
            }
        }
        return foundIndexes;
    }

    private boolean substringIsAnagram (String s, String p) {
        char[] pArr = p.toCharArray();
        for (int c = 0; c < s.length(); c++) {
            int letterIndex = findIndex(pArr, s.charAt(c));
            if (letterIndex > -1) {
                pArr[letterIndex] = ' ';
            } else {
                return false;
            }
        }
        return true;
    }

    private int findIndex(char[] pArr, char c) {
        for (int i = 0; i < pArr.length; i++) {
            if (pArr[i] == c) {
                return i;
            }
        }
        return -1;
    }
}
